### PR TITLE
Milestone #1 -user -interface

### PR DESCRIPTION
**Pull request by Elukoye**
Kindly review game logic

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  return true if (0..2).all? { |i| check(1, i, player) }
+  return true if (0..2).all? { |i| check(2, i, player) }
+  #vertical wins
+  return true if (0..2).all? { |i| check(i, 0, player) }
+  return true if (0..2).all? { |i| check(i, 1, player) }
+  return true if (0..2).all? { |i| check(i, 2, player) }
+  #diagonal wins
+  return true if (0..2).all? { |i| check(i, i, player) }
+  return true if (0..2).all? { |i| check(i, 2 - i, player) }
+  end
```
Comment (maelfosso) [inline main.rb]: Could you indent your code please?

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+end
```
Comment (maelfosso) [inline main.rb]: You should put the class into a separate file

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+player = ["x","o"]
+game.show_instructions
+loop do
+  puts "Its player #{player[0]}'s turn'"
+  input = gets.chomp
+  game.move input, player[0]
+  game.show
+  break if game.win player[0]
+  player[0], player[1] = player[1], player[0]
+end
```
Comment (maelfosso) [inline main.rb]: Loop is :+1: 
Detecting winning position :+1: 
Good way to switch the player

Comment (maelfosso) [reply main.rb]: - The game instructions/loop should not be into the same file than the logic class. Check the project instructions about files structures
- You should say whether or not it's a draw
- Avoid playing in a place already used

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello,

Some suggestions, please
1. You should separate the Game logic with the Board logic. 

**Board**
- The 9 cases
- You check the input. Is it valid or not?
- You check whether or not the place is empty or busy, that is, is there already `:X` or `:O` into that place?
- Is the board in a winning state or not? 
- So on and forth

**Game**
- Start the game
- During the game, I switch the player
- Should I continue or not?
- so on and forth

It's just some suggestions but you should have at least 2 classes implementing specific logic.

2- Make changes on your file structure. The executable file should be `bin/main.rb`

3- After starting playing, could it be possible to help the player to remember the position? **`OPTIONAL`**.

Kindly makes those changes and submit a new code review request.

**Comment by Elukoye** [comment]
Added bin and lib directory to project and added board logic to board.rb. Kindly review

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello,

Nice improvement. You still need to change a few things

- The executable file should be `bin/main.rb`. It's the only file that contains user interaction commands 
- Why does the `players.rb` file is empty?
- Where are the functions which checked whether or not the board has a winning position?
- Just as a reminder: Your project should have at least two classes (Usually it uses three: Game, Board and Player). The code of the project should be logically divided between classes.

Thanks

**Comment by Elukoye** [comment]
Hi mael,
I am working with the milestones so I did the board logic first ,I created
the empty file because its the next mile stone I am working on ... I think
the project says that you need to ask for a code review after every mile
stone.

**Comment by maelfosso** [comment]
Hello @Elukoye 

You're right about the milestone. The thing is, you start proposing something in your first milestone. The reviewer suggests you do some change but what you come back with has nothing to do with your first proposition, whether in execution or content.

Thanks

**Comment by Elukoye** [comment]
Ok understood

**Comment by Elukoye** [comment]
Added game logic to  main.rb  in the bin directory and board logic to board.rb in the lib directory.Kindly review the project.

File: bin/main.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
+require '../lib/board.rb'
```
Comment (maelfosso) [inline bin/main.rb]: Please use `require_relative`

File: lib/board.rb [hunk 5]
```diff
@@ -1,38 +1,24 @@
...
-  #lays out the winning combinations 
-  def winning_combinations
-    #vertical_win
-    @vertical_win = [[1,2,3],[4,5,6],[7,8,9]]
-    #horizontal_win
-    @horizontal_win = [[1,4,7],[2,5,8],[3,6,9]]
-    #diagonal_win
-    @diagonal_win = [[1,5,9],[3,5,7]]
-  end 
-end
```
Comment (maelfosso) [inline lib/board.rb]: Please why did you remove it from here ?

File: bin/main.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  postion.between?(0,8) && !position_taken? 
+end
+
+def switch_player
+  if $current_player == "Player x"
+    $current_player = "Player o "
+  else 
+    $current_player = "Player x"
+  end 
+end
```
Comment (maelfosso) [inline bin/main.rb]: Don't use **Global Variable**

You could use a class called `Game` with a attribute called `@current_player` and here just call `game.switch_player`

File: bin/main.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+$current_player = "Player x"
+$win_combination = [
+  #vertical
+ [0,1,2],[3,4,5],[6,7,8],
+ #horizontal
+ [0,3,6],[1,4,7],[2,5,8],
+ #diagonal
+ [0,4,8],[2,4,6]
+]
+
```
Comment (maelfosso) [inline bin/main.rb]: No **Global Variables** !!!

I suggest you to have 2 class, 
**Board** : where you could put `@win_combination`
**Game** : Where you could have `@winner` and `@current_player`

File: bin/main.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+def turn
+  puts "#{$current_player}" 
+  position = gets.chomp.to_i - 1 
+  if $current_player == "Player x" 
+    @board[position] = "x"
+  else
+    @board[position] = "o" 
+  end
+end
+
```
Comment (maelfosso) [inline bin/main.rb]: Can be a method of `Game` class. You could rewrite it like this

```ruby
def turn(position)
  if (@current_player == "Player x") # @current_player == 1
    @board[position] = "x"
  else
    @board[position] = "o"
  end
end
```

File: bin/main.rb [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+def game_over?
+  win? || draw?
+end
+
```
Comment (maelfosso) [inline bin/main.rb]: Can be a method of the `Game` class rewrite like this for example

```ruby
def game_over?
  @board.win? || @board.draw?
end
```

File: bin/main.rb [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  $win_combination.each{ |element|
+   brian = [@board[element[0]],@board[element[1]],@board[element[2]]].all? { |i| i == "x" } || [@board[element[0]],@board[element[1]],@board[element[2]]].all? { |i| i == "o" }
+   if brian
+    $winner = @board[element[0]] 
+   end
+    break  if brian
+  }
+  
+    return brian
+ end
```
Comment (maelfosso) [inline bin/main.rb]: You can move in into `Board` class

File: bin/main.rb [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+def draw?
+   full? && !win?
+end
+
```
Comment (maelfosso) [inline bin/main.rb]: You can move it into `Board` class

File: bin/main.rb [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+def switch_player
+  if $current_player == "Player x"
+    $current_player = "Player o "
+  else 
+    $current_player = "Player x"
+  end 
+end
+
```
Comment (maelfosso) [inline bin/main.rb]: You can move it into `Game` class.

File: bin/main.rb [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+def position_taken?
+   @board[position] == "x" || @board[position] == "o"
+end
+
+def valid_move?
+  postion.between?(0,8) && !position_taken? 
+end
+
```
Comment (maelfosso) [inline bin/main.rb]: `full?` , `position_taken?`, `valid_move?` can be method of `Board` class and you just call them from `Game` class to check it when receiving an input.

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @Elukoye 

Your game is working. No problem with that.
Now the only thing to do is to put methods into two class for example `Game` and `Board`. Please read the comment for some suggestion.

Thanks.
Don't hesitate to write me on Slack, if you have any question.

File: bin/main.rb (hunk 4 above)
Comment (Elukoye) [reply bin/main.rb]: Done

File: bin/main.rb (hunk 10 above)
Comment (Elukoye) [reply bin/main.rb]: Done

File: bin/main.rb (hunk 7 above)
Comment (Elukoye) [reply bin/main.rb]: done

File: bin/main.rb (hunk 8 above)
Comment (Elukoye) [reply bin/main.rb]: done

**Comment by Elukoye** [comment]
Hey Maelfosso,
I have structured the project as you requested but I am getting trouble
with the game_loop method. It gives me the following error;
logic.rb:54:in `<main>': undefined method `game_loop' for
#<Board:0x0000000002e764c0> (NoMethodError)
I am calling the game_loop method on an object created from the board class.
I have tried creating an object from the GameLogic class but it gives me an
error message of undefined win? method for the object created from the
GameLogic class. Unfortunately, I don't understand very well why I am
getting this error. I will send a pull request on GitHub so you can take a
look and guide me.
Regards,
Elukoye.

File: bin/main.rb [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+class GameLogic
+  def initialize
+    @current_player ="Player x"
+    @winner = nil
+  end 
+
+  def turn
+    puts "#{@current_player}" 
+    position = gets.chomp.to_i - 1 
+    if @current_player == "Player x"
```
Comment (mcrd25) [inline bin/main.rb]: You may shorten this if like following

board[position] = @current_player == 'Player x' ?  'x' : 'o' [OPTIONAL]

File: bin/main.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+      @board[position] = "o" 
+    end
+  end 
+
+  def game_over? 
+    @boar.win? || @board.draw?
+  end
+
+  
+  def switch_player
```
Comment (mcrd25) [inline bin/main.rb]: This could also be shortened as the previous suggestion [OPTIONAL]

File: bin/main.rb [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  end 
+
+  def game_over? 
+    @boar.win? || @board.draw?
+  end
+
+  
+  def switch_player 
+    if @current_player == "Player x"
+       @current_player = "Player o "
```
Comment (mcrd25) [inline bin/main.rb]: This you meant "Player o"

File: lib/board.rb [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    @win_combination = [[0,1,2],[3,4,5],[6,7,8],[0,3,6],[1,4,7],[2,5,8],[0,4,8],[2,4,6]]
+  end
+
+  def display
+    puts "| #{@board[0]} | #{@board[1]} | #{@board[2]} |"
+    puts "| #{@board[3]} | #{@board[4]} | #{@board[5]} |"
+    puts "| #{@board[6]} | #{@board[7]} | #{@board[8]} |"  
+  end
+
+  def instruction
```
Comment (mcrd25) [inline lib/board.rb]: This should not be part of board class, you really ought to have this in the game's main script. [REQUIRED]
No puts statement should be in Game's class logic

File: lib/board.rb [hunk 18]
```diff
@@ -0,0 +1,59 @@
+class Board
+  attr_accessor :board, :win_combination
+  def initialize
+    @board = [1,2,3,4,5,6,7,8,9]
+    @win_combination = [[0,1,2],[3,4,5],[6,7,8],[0,3,6],[1,4,7],[2,5,8],[0,4,8],[2,4,6]]
+  end
+
+  def display
```
Comment (mcrd25) [inline lib/board.rb]: maybe you can simply return the board. and display the board in main script something like.
`puts board.display`

File: bin/main.rb [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+      puts " Damn ,its a draw!"
+      puts " PLAY AGAIN?"
+     end
+  end
+end
+
+game = Board.new
+game.instruction
+game.display 
+game.game_loop
```
Comment (mcrd25) [inline bin/main.rb]: This causes an error because. Please address

File: lib/board.rb [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+  def position_taken? #class game
+    @board[position] == "x" || @board[position] == "o"
+  end
+  
+  def valid? #class game
+    if position.between?(0,8)
+      return true
+    else 
+      puts "Invalid move or input."
```
Comment (mcrd25) [inline lib/board.rb]: Should always try and return something as opposed to outputting it when dealing with class methods or any project class

Comment (mcrd25) [reply lib/board.rb]: Leave `puts` and chomps statements to script files.

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Milestone 1 Review: User Interface with Mock Logic 

Hi @Elukoye and @IBUNHABIBU 👋,

Great attempt so far! There seems to be some sort of mixup with milestones. This is the first milestone so I am going to be reviewing things related to this milestone so that you may move unto the next milestone as soon as possible. Since you've already sort of begun work on next milestones I have comments for those as well.

##REQUIRED CHANGES
- When I run `ruby bin/main.rb` it gives a **No Method Error**. Please address
 Cannot properly test CLI interface until this error is corrected
- I am aware that you know this is only the UI milestone but I do hope you know you must have at least two classes. you currently have a file names `board.rb` with Board class. You have another class called `GameLogic` bit it is located in the main.rb file which should not be the case. It would be more appropriate (When you've reached that point in milestones) that you add that to a game.rb file and create another Player class which stores the player name and player symbol (I suggest you use an actual `:symbol` as opposed to string to store player the symbol as this is immutable and better reflects the immutability of the player symbol. 
Also, you must put your game.rb file into lib, not in the script file. YOU MUST SEPARATE game logic from the script. Game Logic should be inside of classes located within files in lib folder. Hope this helps 😅 




## OPTIONAL CHANGES
- You currently have a player as 'Player x' or 'Player o' but it really should be whatever the user wants to be. You should prompt them for their names.
- In general, I really do think there is room for improvement in how you lay your project out. If you like, we can have a chat on how to improve y'all project. 🙂 
- I also strongly suggest you take the advice on suggestions of the previous reviewer. It is only to make you guys better ruby developers.  Ifor one would be glad to have a reviewer take that much time to give me advice. 😄 

GREAT ATTEMPT SO FAR! LET'S GET IT ON NEXT TRY 💪.
HOPE THIS REVIEW HELPED YOU! 

***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/)

**Comment by Elukoye** [comment]
I have made some changes to the project kindly review

**Comment by Elukoye** [comment]
This branch contains the players class which assigns players and tokens

File: lib/players.rb [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+  attr_accessor :name, :piece
+  def initialize
+    @name = name
+    @piece = piece  
+  end
+ 
+  def start_game
+    puts "Welcome to TIC-TAC-TOE"
+   loop do
+    puts "Input name: "
```
Comment (bolah2009) [inline lib/players.rb]: #### `REQUIRED`

Let's move UI away from game logic.

> `main.rb` file should be responsible for user interaction. This is the only one place where `puts` and `get.chomp` can be used.

File: lib/players.rb [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+    @name = gets.chomp.capitalize
+    puts "choose x or o"
+    @piece =gets.chomp.downcase
```
Comment (bolah2009) [inline lib/players.rb]: #### `REQUIRED`

Let's move UI away from game logic.

> `main.rb` file should be responsible for user interaction. This is the only one place where `puts` and `get.chomp` can be used.

File: lib/players.rb [hunk 23]
```diff
@@ -0,0 +1,29 @@
+class Players
+  attr_accessor :name, :piece
+  def initialize
+    @name = name
+    @piece = piece  
+  end
+ 
+  def start_game
```
Comment (bolah2009) [inline lib/players.rb]: #### `HINT`

:bulb: Let's move out this method to `#bin/main.rb` file. This will satisfy the requirement for the next milestone (**Milestone 2: User interface**). :100: 

```ruby
#bin/main.rb


puts 'Welcome to TIC-TAC-TOE'
loop do
  puts 'Input name: '
  @name = gets.chomp.capitalize
  puts 'choose x or o'
  @piece = gets.chomp.downcase

  value = %w[x o]
  if @piece == value[0]
    puts "#{@name} is #{@piece}"
  elsif @piece == value[1]
    puts "#{@name} is #{@piece}"
    puts 'Lets play'
    break
  else
    puts 'Invalid choice, choose X OR O!'
  end
end
```
#### Expected Output behavior after moving to `bin/main.rb`

```ruby
$ ruby bin/main.rb
Welcome to TIC-TAC-TOE
Input name: 
Bola
choose x or o
X
Bola is x
Input name: 
Ahmed
choose x or o
O
Ahmed is o
Lets play
```

**Review by bolah2009 — CHANGES_REQUESTED** [review]
# Hi Team :wave: 

This is **Milestone 1: Project Setup**. Please check [**Pathwright**](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/) for more details.

## This PR currently lacks the required files for it to move to Milestone 2. :relaxed: 

### Make these two changes. :point_down: 
- Add linter files [see comment](https://github.com/Elukoye/tic-tac-toe/pull/1#pullrequestreview-280672279)
   > [Set up a code linter](https://github.com/microverseinc/linters-config/tree/master/ruby).
- Add the two directories `lib` and `bin` where `bin` contains the `main.rb` file.


> ### Milestone 1: Project setup
> 
> Setup project as it is described in requirements.
> 
> In addition, create 2 main directories: - `lib` - for the game logic - `bin` - with the executable file called `main.rb`.


#### Rename the PR (see image below) and submit for another code review. 

rename PR

### Check [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/) for more details.

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

**Review by bolah2009 — COMMENTED** [review]
#### `REQUIRED`

Add `.stickler.yml` file to the root of the project. Also let's use [this file](https://github.com/microverseinc/linters-config/blob/master/ruby/.stickler.yml). :point_down: 

``` yml
linters:
  rubocop:
    display_cop_names: true
files:
  ignore:
    - "bin/*"
    - "db/*"
    - "config/*"
    - "Guardfile"
    - "Rakefile"
```
- Also, add the `.rubocop.yml` file, this can be found [here](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml)

> Remember that `.stickler.yml` and `.rubocop.yml` file names start with a **dot**

If you are having an issue with setting up linter, read the `Troubleshooting` section of [this repo](https://github.com/microverseinc/linters-config/tree/master/ruby#troubleshooting).

### You can run the linter locally:  💥 - run `rubocop` in your local env:
>     - add `gem 'rubocop'` to `Gemfile` (not sure how to use Gemfile? Read [this](https://bundler.io/v1.15/guides/bundler_setup.html))
>     - run `bundle install`
>     - copy [.rubocop.yml](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml) to the root directory of your project
>     - run `rubocop`
>     - fix linter errors
>     - **Remember to let your Code Reviewer know that you had problems with Stickler and you used linter in local env.**

**Comment by IBUNHABIBU** [comment]
Modified user-interface-branch 
kindly review

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @Elukoye @IBUNHABIBU 

Sorry to bother you, but please could you explain us why this PR for ?

There are a lot of suggestion about what should be done for *Milestone 1* and what should be move on *Milestone 2*.
You've modified the `user-interface` branch and your code is very very very good now.
Now in your `user-interface` branch there are  
- `lib` folder, 
- `stickler.yml` 
- README

To be validated, i means *Milestone 1*, you need to setup the project with

- `bin` folder
- `bin/main.rb` folder
- `lib` folder
- `README` file
- Setup `stickler.yml`/
-

Except those point listed above, all the others suggestions are about what need to be move for the *Milestone 2* as @mcrd25 and @bolah2009  said in comment.

Kindly re-organize your code as suggested above for the *Milestone 1* to be approved.
Thanks
Don't hesitate to write  on Slack to make the *Milestone 1* being validated and then the *Milestone 2*, because there is almost everything to over the whole project.
